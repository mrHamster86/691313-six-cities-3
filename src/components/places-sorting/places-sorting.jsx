import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {OFFERS_SORT_ITEMS} from '../../constatnts';
import withBooleanState from '../../hocs/with-boolean-state/with-boolean-state';
import {ActionCreator} from '../../reducer/offers/reducer';
import NameSpace from '../../reducer/name-space';

export const PlacesSorting = ({currentSort, changeSort, booleanState, onToggle}) => {
  return (
    <form
      className="places__sorting"
      action="#"
      method="get"
      onClick={onToggle}
    >
      <span className="places__sorting-caption">Sort by</span>
      <span className="places__sorting-type" tabIndex="0">
        {currentSort}
        <svg className="places__sorting-arrow" width="7" height="4">
          <use xlinkHref="#icon-arrow-select"/>
        </svg>
      </span>
      <ul className={`places__options places__options--custom ${booleanState && `places__options--opened`}`}>
        {OFFERS_SORT_ITEMS.map((sort) => (
          <li
            key={sort}
            className={`places__option ${sort === currentSort && `places__option--active`}`}
            tabIndex="0"
            onClick={changeSort.bind({}, sort)}
          >{sort}</li>
        ))}
      </ul>
    </form>
  );
};

const mapStateToProps = (state) => ({
  currentSort: state[NameSpace.OFFERS].sort
});

const mapDispatchToProps = {
  changeSort: (sort) => ActionCreator.changeSort(sort)
};

PlacesSorting.propTypes = {
  currentSort: PropTypes.string.isRequired,
  changeSort: PropTypes.func.isRequired,
  booleanState: PropTypes.bool.isRequired,
  onToggle: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(withBooleanState(PlacesSorting));
