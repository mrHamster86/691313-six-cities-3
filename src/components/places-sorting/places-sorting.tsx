import * as React from 'react';
import {connect} from 'react-redux';
import {OFFERS_SORT_ITEMS} from '../../constatnts';
import withBooleanState from '../../hocs/with-boolean-state/with-boolean-state';
import {ActionCreator} from '../../reducer/offers/offers';
import {getSort} from '../../reducer/offers/selectors';

interface Props {
  currentSort: string;
  onChangeSort: () => void;
  booleanState: boolean;
  onToggle: () => void;
}

export const PlacesSorting: React.FC<Props> = ({currentSort, onChangeSort, booleanState, onToggle}: Props) => {
  return (
    <form
      className="places__sorting"
      action="#"
      method="get"
      onClick={onToggle}
    >
      <span className="places__sorting-caption">Sort by</span>
      <span className="places__sorting-type">
        {currentSort}
        <svg className="places__sorting-arrow" width="7" height="4">
          <use xlinkHref="#icon-arrow-select"/>
        </svg>
      </span>
      <ul className={`places__options places__options--custom ${booleanState && `places__options--opened`}`}>
        {OFFERS_SORT_ITEMS.map((sort) => (
          <li
            key={sort}
            className={`places__option ${sort === currentSort && `places__option--active`}`}
            onClick={onChangeSort.bind({}, sort)}
          >{sort}</li>
        ))}
      </ul>
    </form>
  );
};

const mapStateToProps = (state) => ({
  currentSort: getSort(state)
});

const mapDispatchToProps = {
  onChangeSort: (sort) => ActionCreator.changeSort(sort)
};

export default connect(mapStateToProps, mapDispatchToProps)(withBooleanState(PlacesSorting));
