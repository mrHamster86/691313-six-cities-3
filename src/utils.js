export const extend = (a, b) => {
  return Object.assign({}, a, b);
};

export const updateObjectInArray = (array, newObject, key) => {
  const newArray = [...array];
  const index = newArray.findIndex((item) => item[key] === newObject[key]);
  newArray[index] = extend(newArray[index], newObject);
  return newArray;
};

export const noop = () => {
  // do nothing
};
