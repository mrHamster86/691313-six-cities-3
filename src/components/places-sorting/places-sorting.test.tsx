import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {PlacesSorting} from './places-sorting';
import {noop} from '../../utils';

const currentSort = `Popular`;
const booleanState = false;

it(`PlacesSorting components renders correctly`, () => {
  const tree = renderer.
    create(<PlacesSorting
      currentSort={currentSort}
      onChangeSort={noop}
      booleanState={booleanState}
      onToggle={noop}
    />).
    toJSON();
  expect(tree).toMatchSnapshot();
});
