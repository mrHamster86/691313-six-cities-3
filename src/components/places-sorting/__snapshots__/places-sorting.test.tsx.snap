// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlacesSorting components renders correctly 1`] = `
<form
  action="#"
  className="places__sorting"
  method="get"
  onClick={[Function]}
>
  <span
    className="places__sorting-caption"
  >
    Sort by
  </span>
  <span
    className="places__sorting-type"
  >
    Popular
    <svg
      className="places__sorting-arrow"
      height="4"
      width="7"
    >
      <use
        xlinkHref="#icon-arrow-select"
      />
    </svg>
  </span>
  <ul
    className="places__options places__options--custom false"
  >
    <li
      className="places__option places__option--active"
      onClick={[Function]}
    >
      Popular
    </li>
    <li
      className="places__option false"
      onClick={[Function]}
    >
      Price: low to high
    </li>
    <li
      className="places__option false"
      onClick={[Function]}
    >
      Price: high to low
    </li>
    <li
      className="places__option false"
      onClick={[Function]}
    >
      Top rated first
    </li>
  </ul>
</form>
`;
